### ISSUE TITLE
why isn't string multiplication is not commutative?

### ISSUE THREAD

**Issue by ceremcem** [gkz/LiveScript#740]
why isn't `4 * 'a'` equal to `'a' * 4`?

**Comment by phanimahesh** [comment]
I don't suppose this is by design? The empty string makes no sense, so I think this is a bug.

Also, fun fact: `'a' * 'b' == 'a'`. Another bug?

**Comment by gabeio** [comment]
@phanimahesh what would you want `'a' * 'b'` to equal? in node it equals `NaN`...

**Comment by ceremcem** [comment]
Yes it seems to be another bug. I think `'a' * 'b'` should throw an exception (or like) because it is meaningless.

**Comment by gabeio** [comment]
If node doesn't throw an error why throw an error?

**Comment by phanimahesh** [comment]
It would be good if LS fixed the warts of js. Anything other than a syntax
error doesn't make sense to me.

Why? Because this can only be a typo, and makes absolutely no sense
whatsoever. At least to me. And telling the developer that they messed up
is always good. Fail early to prevent lots of head scratching and
frustration later.

**Comment by igl** [comment]
You can still do `a = 'a'; a * 3` and you can't fetch that without type checking.

Fetching warts sounds nice but if you can't go all the way anyway, why bother making the parser complicated? Sounds more like a job for a linter anyway.

**Comment by satyr** [comment]
`whatever * 'string'` is short for `Array.prototype.join.call(whatever, 'string')`. Those results are expected.

```
$ node
> [].join.call(4, 'a')
''
> [].join.call('a', 'b')
'a'
> [].join.call('abc', '_')
'a_b_c'
```

**Comment by phanimahesh** [comment]
Oh, that makes sense. And yes, I can see that we can't tackle this problem
without type checking.

We can file this under warts of LS and let things be. A wiki page maybe?

**Comment by gabeio** [comment]
I feel like I am missing something. why does `* == join` ?

**Comment by ceremcem** [comment]
LS should throw an exception (or like) even if node.js doesn't in the same conditions because "*LiveScript is a language* that compiles to Javascript". It's another language. This isn't bad.

**Comment by phanimahesh** [comment]
The problem is that it is not possible to detect this error fully without having a type system, declared or inferred. Without that, only the most trivial case of a two literal strings can be caught as an error, and that is not worth the added complexity in parser. Also, that would make LiveScript inconsistent, and we'd get another issue along the lines of

    'a' * 'b' # Error
    ---
    a = 'a'
    a * 'b' # Not an error.

**Comment by impinball** [comment]
Okay...that is a good catch. @vendethiel PTAL...I can repro this, can you?

**Comment by impinball** [comment]
And `'a' * 'b' == 'a'` is surprising...it should throw. `'a'.repeat('b')` throws in ES2015 and most utility libraries that have this.

**Comment by zloirock** [comment]
@impinball [`'a'.repeat('b') -> ''`](http://www.ecma-international.org/ecma-262/6.0/#sec-string.prototype.repeat)

**Comment by impinball** [comment]
Well...I'll take that back, then. I am completely and utterly confused, then (it should be an easy fix, though).

**Comment by impinball** [comment]
And what it compiles to is (roughly) this:

```ls
[].join.call \a \b
